client/state/posts/reducer.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import indexBy from 'lodash/collection/indexBy';
+
+/**
+ * Internal dependencies
+ */
+import {
+	POST_RECEIVE,
+	POSTS_RECEIVE,
+	POSTS_REQUEST,
+	POSTS_REQUEST_SUCCESS,
+	POSTS_REQUEST_FAILURE
+} from 'state/action-types';
+
+/**
+ * Tracks all known post objects, indexed by post global ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+			state = Object.assign( {}, state, {
+				[ action.post.global_ID ]: action.post
+			} );
+			break;
+
+		case POSTS_RECEIVE:
+			state = Object.assign( {}, state, indexBy( action.posts, 'global_ID' ) );
+			break;
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated site posts state after an action has been dispatched.
+ * The state reflects a mapping of site ID, post ID pairing to global post ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function sitePosts( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+		case POSTS_RECEIVE:
+			let posts = action.post ? [ action.post ] : action.posts;
+			posts.forEach( ( post ) => {
+				state = Object.assign( {}, state );
+				if ( ! state[ post.site_ID ] ) {
+					state[ post.site_ID ] = {};
+				}
+
+				state[ post.site_ID ][ post.ID ] = post.global_ID;
+			} );
```

Comment:
It looks like we'll end up making lots of clones here and I wonder if you didn't intend on cloning state outside of the `forEach` loop.